### ISSUE TITLE
BUILDPLATFORM ARG empty for image build with docker-compose

### ISSUE THREAD

**Issue by fuchst** [oauth2-proxy/oauth2-proxy#1729]
Building an image with docker-compose fails because the **BUILDPLATFORM** ARG is empry for the call `FROM --platform=${BUILDPLATFORM} golang:1.17-buster AS builder` in the *Dockerfile* although it is specified in the *docker-compose.yml*.

```
ERROR: Service 'oauth2-proxy' failed to build: failed to parse platform : "" is an invalid component of "": platform specifier component must match "^[A-Za-z0-9_-]+$": invalid argument
```

## Expected Behavior

An image of oauth2-proxy should be build.

## Current Behavior

It fails because **BUILDPLATFORM** is empty ("").

## Possible Solution

Move `ARG BUILDPLATFORM` to the top of the *Dockerfile*.

## Steps to Reproduce (for bugs)

1.  Checkout the git repository to *./oauth2-proxy*
2.  Create the docker-compose.yml
  ```
  version: "3"
  services:
    oauth2-proxy:
      build:
        context: ./oauth2-proxy
        dockerfile: Dockerfile
        args:
          VERSION: lucid_heisenberg
          TARGETPLATFORM: linux/amd64
          BUILDPLATFORM: linux/amd64
  ```
3. Build
  ```
  docker-compose build
  ``` 

## Your Environment

Ubuntu 20.04
Docker 20.10.7
Docker-Compose 1.25.0

- Version used: 7a784a460dbccad2fff4837cdad5a9fb414f971c

**Comment by JoelSpeed** [comment]
The build args are in a particular location to improve the caching, that said, moving the `BUILDPLATFORM` arg is not likely to affect the caching detrimentally so it could be done.

Have you tried building using our Makefiles? They should work across platforms though you need a recent version of Docker with buildx support

**Comment by fuchst** [comment]
Thanks for the answer. I looked a bit into `docker buildx` and it cleared the situation.

`docker buildx` sets the **TARGETPLATFORM** and **BUILDPLATFORM** variables automatically and building works.
For normal `docker build` the variable stays empty until the *ARG* command sets it in the Dockerfile.

I also tried the `x-bake` docker-compose extension from `docker buildx`, but it did not set **BUILDPLATFORM** either. 

So with the current placement of **ARG BUILDPLATFORM** it seems using `docker buildx` is the only way to build.

If moving **ARG BUILDPLATFORM** to the top  does not have negative impact on the current build process, I would keep the issue as an enhancement, otherwise it can be closed.

**Comment by glory-ka** [comment]
I had the exact same issue when I tried building an Alpine Linux docker image on Amazon Linux. It seems the `--platform=$BUILDPLATFORM and --platform=$TARGETPLATFORM` flags were recently added to speed up the build. The commit message said: `"Add BUILDPLATFORM and TARGETPLATFORM` flags to speed up builds".
Since I was more interested in testing something on Alpine Linux than the build speed I simply removed these flags. It worked for me.
